db/trie/leafiterator.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2018 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package trie
+
+// Iterator iterates a trie
+type Iterator interface {
+	Next() bool
+	Error() error
+	Current() ([]byte, []byte)
+}
+
+// LeafIterator defines an iterator to go through all the leaves under given node
+type LeafIterator struct {
+	tr    Trie
+	cur   Node
+	err   error
+	stack []Node
+}
+
+// NewLeafIterator returns a new leaf iterator
+func NewLeafIterator(tr Trie) Iterator {
+	stack := []Node{tr.Root()}
+
+	return &LeafIterator{tr: tr, stack: stack}
+}
+
+// Next moves iterator to next node
+func (li *LeafIterator) Next() bool {
```

Comment:
How about following this semantics: https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html? Check will not move the needle, fetch will.